sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala

```diff
@@ -842,6 +842,120 @@ class FileBasedDataSourceSuite extends QueryTest
       }
     }
   }
+
+  test("SPARK-31116: Select nested parquet with case insensitive mode") {
+    Seq("true", "false").foreach { nestedSchemaPruningEnabled =>
+      withSQLConf(
+        SQLConf.CASE_SENSITIVE.key -> "false",
+        SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key -> nestedSchemaPruningEnabled) {
+        withTempPath { dir =>
+          val path = dir.getCanonicalPath
+
+          // Prepare values for testing nested parquet data
+          spark
+            .range(1L)
+            .selectExpr("NAMED_STRUCT('lowercase', id, 'camelCase', id + 1) AS StructColumn")
+            .write.parquet(path)
+
+          val exactSchema = "StructColumn struct<lowercase: LONG, camelCase: LONG>"
+
+          checkAnswer(spark.read.schema(exactSchema).parquet(path), Row(Row(0, 1)))
+
+          // In case insensitive manner, parquet's column cases are ignored
+          val innerColumnCaseInsensitiveSchema =
+            "StructColumn struct<Lowercase: LONG, camelcase: LONG>"
+          checkAnswer(
+            spark.read.schema(innerColumnCaseInsensitiveSchema).parquet(path),
+            Row(Row(0, 1)))
+
+          val rootColumnCaseInsensitiveSchema =
+            "structColumn struct<lowercase: LONG, camelCase: LONG>"
+          checkAnswer(
+            spark.read.schema(rootColumnCaseInsensitiveSchema).parquet(path),
+            Row(Row(0, 1)))
+        }
+      }
+    }
+  }
+
+  test("SPARK-31116: Select simple parquet with case sensitive mode") {
+    Seq("true", "false").foreach { nestedSchemaPruningEnabled =>
+      withSQLConf(
+        SQLConf.CASE_SENSITIVE.key -> "true",
+        SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key -> nestedSchemaPruningEnabled) {
+        withTempPath { dir =>
+          val path = dir.getCanonicalPath
+
+          // Prepare values for testing specific parquet record reader
+          Seq("A").toDF("camelCase").write.parquet(path)
+
+          val exactSchema = new StructType().add("camelCase", StringType)
+          checkAnswer(spark.read.schema(exactSchema).parquet(path), Row("A"))
+
+          // In case sensitive manner, different letter case does not read column
+          val caseInsensitiveSchema = new StructType().add("camelcase", StringType)
+          checkAnswer(spark.read.schema(caseInsensitiveSchema).parquet(path), Row(null))
+
+          // It also properly work in combined schema
```

Comment:
Again, there is no issue even in case-sensitive manner with non-nested schema in Apache Spark master.